### PR TITLE
Akka http server responds HEAD request as GET

### PR DESCRIPTION
**Pull request by SugandhaAgrawal**
Akka http client by default treats any http HEAD request as GET unless the configuration sets it otherwise. There are use cases where user expects HEAD to be responded as HEAD and without a response body unlike GET.

https://github.com/akka/akka-http/issues/2088 mentions that the default is set to treat a HEAD as a GET request by the Akka client.

## Description
Update the controller configuration to process HEAD and GET requests as they are expected.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by chetanmeh** [comment]
@SugandhaAgrawal I believe this is Akka HTTP Server setting and not related to client. 

If we enable this would HEAD request still work in OpenWhisk or would fail. I presume so far they work just that internally executed as GET. However with this change would they stop working if route does not implement HEAD support?

**Comment by markusthoemmes** [comment]
It'd be useful to get some context on the use-cases that trigger this issue.

**Comment by mbehrendt** [comment]
the context of this issue is actually quite straightforward: We have seen users trying to implement a web action that responds differently on HEAD than on GET. However, given `__ow_method` always had the same value (`get`), they weren't able to make that distinction in their action code.

**Comment by markusthoemmes** [comment]
Got it, that makes sense. I guess @chetanmeh was confused (so was I) because the initial description talks about akka-https client not behaving as expected.

@chetanmeh is also right though that this is a breaking change as currently working HEAD requests will now fail.

**Comment by mbehrendt** [comment]
i wouldn't consider to be a breaking change per se, but rather a bug fix. It was just flatout wrong that when a `HEAD` request came in, `__ow_method` showed `GET`. Of course, if people relied on that bug to be true forever, they would break, but that is true for probably most other bug fixes as well :-)

**Comment by SugandhaAgrawal** [comment]
@markusthoemmes I have the same thought as Michael's. I would say this is a bug fix. The breaking change in terms that the users need to change the implementation is fine but this is right way to handle the HEAD request.

**Comment by SugandhaAgrawal** [comment]
@markusthoemmes The initial description does mean that akka http client isnt behaving as expected *by default*. It responds HEAD with a GET and that isnt the right response in my opinion. 
In order to achieve this you need to set explicitly the value to off. My expectation was that the default should have been off or do you not find this as an expected result and correct?

**Comment by markusthoemmes** [comment]
@SugandhaAgrawal the point is that it's the akka http **server**, not the client.

I'm indecisive on the bug fix vs. breaking change. I agree though that the amount of breakage is very small and arguably relying on undocumented/wrong behavior.

If @chetanmeh doesn't have strong feelings against it I'm fine merging this as-is. Otherwise we should take it to the dev-list.

**Comment by SugandhaAgrawal** [comment]
@markusthoemmes Thanks for the correction, server ;)

**Comment by SugandhaAgrawal** [comment]
@chetanmeh I will do include test in the next PR

**Comment by chetanmeh** [comment]
If this fix is not very urgent would it be possible to include the test in this PR itself. I believe you can still enable this flag in your setups via system property

**Comment by rabbah** [comment]
+1 to include a test.

**Comment by rabbah** [comment]
I agree this is a bug - the intended behavior is for the HEAD to reach the function just like OPTIONS. You'll need a _system_ level tests that "curls" the web action. I tested this patch locally and don't see a difference in behavior (a GET is posted with and without the patch). So +1 from me to include a test in this PR.

**Comment by rabbah** [comment]
@chetanmeh I'm not sure if existing actions would fail --- at least not all. I suspect that you need actions which have already opted out of the default CORS response.

**Comment by rabbah** [comment]
@SugandhaAgrawal any luck adding a test for this fix?

**Comment by SugandhaAgrawal** [comment]
@rabbah Unfortunately not. I was on vacation and returned today. Will work on it

**Comment by rabbah** [comment]
welcome back!

File: tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala [hunk 1]
```diff
@@ -305,6 +305,34 @@ class WskWebActionsTests extends TestHelpers with WskTestHelpers with RestUtil w
...
   }
 
+  /**
+   * Tests web action for HEAD requests
+   */
+  it should "create a web action making a HEAD request" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
+    val name = "webaction"
+    val file = Some(TestUtils.getTestActionFilename("echo.js"))
+    val host = getServiceURL()
+    val url = s"$host$testRoutePath/$namespace/default/$name.text/__ow_user"
```
Comment (rabbah) [inline tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala]: you should not need this here for a HEAD test: `.text/__ow_user`

File: tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala [hunk 2]
```diff
@@ -305,6 +305,34 @@ class WskWebActionsTests extends TestHelpers with WskTestHelpers with RestUtil w
...
+    val authorizedResponse = RestAssured
+      .given()
+      .config(sslconfig)
+      .auth()
+      .preemptive()
+      .basic(wskprops.authKey.split(":")(0), wskprops.authKey.split(":")(1))
+      .head(url)
+
+    authorizedResponse.statusCode shouldBe 200
+    authorizedResponse.body.asString() shouldBe ""
```
Comment (rabbah) [inline tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala]: you'll want to test that the `__ow_method` received was HEAD not GET.

**Comment by SugandhaAgrawal** [comment]
Screenshot 2019-06-27 at 09 45 52


Travis System Test

File: tests/dat/actions/echo-web-http-head.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more contributor
+// license agreements; and to You under the Apache License, Version 2.0.
```
Comment (rabbah) [inline tests/dat/actions/echo-web-http-head.js]: Please use the long form license. We have to stop using the short form.

Comment (rabbah) [reply tests/dat/actions/echo-web-http-head.js]: it was a recent change.

thanks for the update!

File: tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala [hunk 4]
```diff
@@ -305,6 +305,35 @@ class WskWebActionsTests extends TestHelpers with WskTestHelpers with RestUtil w
...
+   * Tests web action for HEAD requests
+   */
+  it should "create a web action making a HEAD request" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
+    val name = "webactionHead"
+    val file = Some(TestUtils.getTestActionFilename("echo-web-http-head.js"))
+    val host = getServiceURL()
+    val url = s"$host$testRoutePath/$namespace/default/$name"
+
+    assetHelper.withCleaner(wsk.action, name) { (action, _) =>
+      action.create(name, file, web = Some("true"), annotations = Map("require-whisk-auth" -> true.toJson))
```
Comment (rabbah) [inline tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala]: i think you can skip the require-whisk-auth annotation here.

File: tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala [hunk 5]
```diff
@@ -305,6 +305,35 @@ class WskWebActionsTests extends TestHelpers with WskTestHelpers with RestUtil w
...
+    }
+
+    val unauthorizedResponse = RestAssured.given().config(sslconfig).head(url)
+    unauthorizedResponse.statusCode shouldBe 401
```
Comment (rabbah) [inline tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala]: and then remove this unless this is important to the test.

Comment (SugandhaAgrawal) [reply tests/src/test/scala/org/apache/openwhisk/core/cli/test/WskWebActionsTests.scala]: It is as a pattern also used for the other web action tests above. Hence, I included it. The behaviour is not explicit for head or get request so I can remove it

**Comment by SugandhaAgrawal** [comment]
@rabbah Please merge this, if everything looks good!

**Comment by rabbah** [comment]
Thanks @SugandhaAgrawal 👍

**Pull request merged**